### PR TITLE
monitor: Refactor listener registration logic

### PR DESCRIPTION
**Pull request by michi-covalent**
- Modify registerNewListener() to take MonitorListener as a parameter
  to allow arbitrary listener to be registered instead of assuming the
  type of listener is always listenerv1_2.
- Add Close() method to MonitorListener so that the Monitor can close
  listeners without knowing implementation details.

Ref #9925

Signed-off-by: Michi Mutsuzaki <michi@isovalent.com>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9924)

### CODE REVIEW

File: pkg/monitor/agent/listener1_2.go
Comment (joestringer) [inline pkg/monitor/agent/listener1_2.go]: And what? Don't leave me hanging here michi

File: pkg/monitor/agent/listener1_2.go
Comment (joestringer) [inline pkg/monitor/agent/listener1_2.go]: Interesting, were we leaking the ml.queue channel here before? I wonder if we need a separate bugfix for older releases.

**Review by joestringer — APPROVED** [review]
Minor nits below, lgtm.

**Comment by joestringer** [comment]
test-me-please

File: pkg/monitor/agent/listener1_2.go
Comment (michi-covalent) [reply pkg/monitor/agent/listener1_2.go]: got distracted while writing this comment :) updating...

File: pkg/monitor/agent/listener1_2.go
Comment (michi-covalent) [reply pkg/monitor/agent/listener1_2.go]: yes it does look like ml.queue was never closed. i can split this part into a separate PR if it makes it easier to backport.

Comment (joestringer) [reply pkg/monitor/agent/listener1_2.go]: @michi-covalent please do. Looks like we'll wait until branching for the next release before we merge this PR anyway.

Comment (michi-covalent) [reply pkg/monitor/agent/listener1_2.go]: on the second thought, looks like go garbage collects unused channels? https://groups.google.com/forum/#!msg/golang-nuts/pZwdYRGxCIk/qpbHxRRPJdUJ

i needed to add `close(ml.queue)` in this pr to unblock `drainQueue` from here: https://github.com/cilium/cilium/blob/c6e876c1ac568db853af21ccde5445cef9da1a5e/pkg/monitor/agent/listener1_2.go#L63

Comment (joestringer) [reply pkg/monitor/agent/listener1_2.go]: OK, in that case don't worry.

**Comment by joestringer** [comment]
test-me-please

**Comment by michi-covalent** [comment]
test-me-please

**Comment by michi-covalent** [comment]
@aanm @joestringer can this be merged now?

File: pkg/monitor/agent/listener1_2.go
Comment (tgraf) [inline pkg/monitor/agent/listener1_2.go:89]: This `close()` call is new and not mentioned in the commit message. Why is it needed?

**Review by tgraf — CHANGES_REQUESTED** [review]

File: pkg/monitor/agent/listener1_2.go
Comment (michi-covalent) [reply pkg/monitor/agent/listener1_2.go:89]: this is added to unblock drainQueue goroutine in unit test:

https://github.com/cilium/cilium/pull/9924#discussion_r369336250

i'll update the commit message

Comment (michi-covalent) [reply pkg/monitor/agent/listener1_2.go:89]: @tgraf i updated the commit message

**Review by tgraf — APPROVED** [review]

**Comment by tgraf** [comment]
test-me-please

**Comment by michi-covalent** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

**Comment by michi-covalent** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Pull request merged**
